scripts/esbuild/testing.ts

```diff
@@ -0,0 +1,141 @@
+import type { BuildOptions as ESBuildOptions, Plugin } from 'esbuild';
+import fs from 'fs-extra';
+import { join } from 'path';
+
+import { copyTestingInternalDts } from '../bundles/testing';
+import { getBanner } from '../utils/banner';
+import type { BuildOptions } from '../utils/options';
+import { writePkgJson } from '../utils/write-pkg-json';
+import { getBaseEsbuildOptions, getEsbuildAliases, getEsbuildExternalModules, runBuilds } from './util';
+
+const EXTERNAL_TESTING_MODULES = [
+  'assert',
+  'buffer',
+  'child_process',
+  'console',
+  'constants',
+  'crypto',
+  'fs',
+  '@jest/core',
+  'jest-cli',
+  'jest',
+  'expect',
+  '@jest/reporters',
+  'jest-environment-node',
+  'jest-message-id',
+  'jest-runner',
+  'net',
+  'os',
+  'path',
+  'process',
+  'puppeteer',
+  'puppeteer-core',
+  'readline',
+  'rollup',
+  '@rollup/plugin-commonjs',
+  '@rollup/plugin-node-resolve',
+  'stream',
+  'tty',
+  'url',
+  'util',
+  'vm',
+  'yargs',
+  'zlib',
+];
+
+export async function buildTesting(opts: BuildOptions) {
+  const inputDir = join(opts.buildDir, 'testing');
+  const sourceDir = join(opts.srcDir, 'testing');
+
+  await Promise.all([
+    // copy jest testing entry files
+    fs.copy(join(opts.scriptsBundlesDir, 'helpers', 'jest'), opts.output.testingDir),
+    copyTestingInternalDts(opts, inputDir),
+  ]);
+
+  // write package.json
+  writePkgJson(opts, opts.output.testingDir, {
+    name: '@stencil/core/testing',
+    description: 'Stencil testing suite.',
+    main: 'index.js',
+    types: 'index.d.ts',
+  });
+
+  const external = [
+    ...EXTERNAL_TESTING_MODULES,
+    ...getEsbuildExternalModules(opts, opts.output.testingDir),
+    '@rollup/plugin-commonjs',
+    '@rollup/plugin-node-resolve',
+    'rollup',
+    '../compiler/stencil.js',
+  ];
+
+  const testingEsbuildOptions: ESBuildOptions = {
+    ...getBaseEsbuildOptions(),
+    entryPoints: [join(sourceDir, 'index.ts')],
+    bundle: true,
+    format: 'cjs',
+    outfile: join(opts.output.testingDir, 'index.js'),
+    platform: 'node',
+    logLevel: 'info',
+    external,
+    write: false,
```

Comment:
just curious, what's the reason for `write: false` here? based on https://esbuild.github.io/api/#write it looks like that should prevent the bundle from being written to disk, but strangely it doesn't seem to do that 🤔

Reply:
We have to do some last modifications on the output code in `lazyRequirePlugin` to enable lazy require to work. I am not sure if there is a better way do to that but it seems like I can either have a plugin that reads the file from disk to modify the source and push it along the pipeline or do it before it is being written to disk.